### PR TITLE
[generator] Cities boundaries builder.

### PR DESCRIPTION
**Pull request by ygorshenin**

### CODE REVIEW

File: generator/cities_boundaries_builder.cpp [hunk 1]
```diff
@@ -0,0 +1,106 @@
...
+};
+
+CBV GetLocalities(string const & dataPath)
+{
+  Index index;
+  auto const result = index.Register(platform::LocalCountryFile::MakeTemporary(dataPath));
+  CHECK_EQUAL(result.second, MwmSet::RegResult::Success, ("Can't register", dataPath));
+
+  search::MwmContext context(index.GetMwmHandleById(result.first));
+  return search::CategoriesCache(LocalitiesSource{}, my::Cancellable{}).Get(context);
```
Comment (mpimenov) [inline generator/cities_boundaries_builder.cpp]: Are you sure there's no use-after-free here, when the cache is destroyed at scope exit?

Comment (mpimenov) [reply generator/cities_boundaries_builder.cpp]: Ah, CategoriesCache::Get() returns by copy. Ok.

Comment (ygorshenin) [reply generator/cities_boundaries_builder.cpp]: Cache exists until the end of the full expression, no UAF in any case.

File: generator/osm_source.cpp [hunk 2]
```diff
@@ -1,3 +1,4 @@
+#include "generator/cities_boundaries_builder.hpp"
```
Comment (mpimenov) [inline generator/osm_source.cpp]: osm_source.hpp should be at the top

File: generator/osm_translator.hpp [hunk 3]
```diff
@@ -374,6 +383,13 @@ class OsmToFeatureTranslator
     // Key point here is that IsDrawableLike and RemoveNoDrawableTypes
     // work a bit different for GEOM_AREA.
 
+    if (IsCityBoundary(params))
```
Comment (mpimenov) [inline generator/osm_translator.hpp]: Please move this above the comment to IsDrawableLike.

File: indexer/indexer_tests/cities_boundaries_serdes_tests.cpp [hunk 4]
```diff
@@ -23,78 +23,88 @@ namespace
 using Boundary = vector<CityBoundary>;
 using Boundaries = vector<Boundary>;
 
-void TestEqual(BoundingBox const & lhs, BoundingBox const & rhs, double eps)
+static_assert(CitiesBoundariesSerDes::kLatestVersion == 0, "");
+static_assert(CitiesBoundariesSerDes::HeaderV0::kDefaultCoordBits == 19, "");
+
+// Precision of mercator coords encoded with 19 bits.
```
Comment (mpimenov) [inline indexer/indexer_tests/cities_boundaries_serdes_tests.cpp]: s/Precision/Absolute precision/

File: indexer/cities_boundaries_serdes.hpp [hunk 5]
```diff
@@ -131,6 +145,14 @@ class CityBoundaryEncoder
       return us;
     }
 
+    void EncodeDelta(m2::PointU const & curr, m2::PointU const & next)
+    {
+      auto const dx = static_cast<int32_t>(next.x) - static_cast<int32_t>(curr.x);
```
Comment (mpimenov) [inline indexer/cities_boundaries_serdes.hpp]: asserted_cast?

File: generator/utils.hpp [hunk 6]
```diff
@@ -1,8 +1,38 @@
...
+#include "coding/reader.hpp"
+
+#include "base/logging.hpp"
+
 namespace generator
 {
 void LoadIndex(Index & index);
+
+template <typename ToDo>
+bool ForEachOsmId2FeatureId(string const & path, ToDo && toDo)
```
Comment (bykoianko) [inline generator/utils.hpp]: `#include <string>`, `std::`?

File: generator/cities_boundaries_builder.cpp [hunk 7]
```diff
@@ -0,0 +1,106 @@
...
+
+using namespace indexer;
+using namespace search;
+using namespace std;
+
+namespace generator
+{
+namespace
+{
+bool ParseFeatureIdToOsmIdMapping(string const & path, map<uint32_t, vector<osm::Id>> & mapping)
```
Comment (bykoianko) [inline generator/cities_boundaries_builder.cpp]: `#include <cstdint>`?

File: generator/utils.hpp (hunk 6 above)
Comment (bykoianko) [reply generator/utils.hpp]: Хорошо, что ты добавил такой метод. Мне такой очень скоро понадобится для transit.

File: indexer/cities_boundaries_serdes.hpp [hunk 8]
```diff
@@ -315,4 +347,100 @@ class CityBoundaryDecoder
...
+    Sink & m_sink;
+  };
+
+  template <typename Source>
+  struct ReadFromSourceVisitor
+  {
+    ReadFromSourceVisitor(Source & source) : m_source(source) {}
+
+    template <typename T>
+    typename enable_if<is_integral<T>::value || is_enum<T>::value, void>::type operator()(
```
Comment (bykoianko) [inline indexer/cities_boundaries_serdes.hpp]: is_class<> вместо is_integral<> || is_enum<>?

File: indexer/indexer_tests/cities_boundaries_serdes_tests.cpp [hunk 9]
```diff
@@ -23,78 +23,88 @@ namespace
...
+double const kEps = 1e-3;
```
Comment (bykoianko) [inline indexer/indexer_tests/cities_boundaries_serdes_tests.cpp]: В TestEqual для PointD не много ли 10^-3 погрешность?

File: generator/cities_boundaries_builder.cpp (hunk 7 above)
Comment (ygorshenin) [reply generator/cities_boundaries_builder.cpp]: Done

File: generator/osm_source.cpp (hunk 2 above)
Comment (ygorshenin) [reply generator/osm_source.cpp]: Done

File: generator/osm_translator.hpp (hunk 3 above)
Comment (ygorshenin) [reply generator/osm_translator.hpp]: Done

File: generator/utils.hpp (hunk 6 above)
Comment (ygorshenin) [reply generator/utils.hpp]: Done

File: indexer/cities_boundaries_serdes.hpp (hunk 5 above)
Comment (ygorshenin) [reply indexer/cities_boundaries_serdes.hpp]: Done

File: indexer/cities_boundaries_serdes.hpp (hunk 8 above)
Comment (ygorshenin) [reply indexer/cities_boundaries_serdes.hpp]: Why?

File: indexer/indexer_tests/cities_boundaries_serdes_tests.cpp (hunk 4 above)
Comment (ygorshenin) [reply indexer/indexer_tests/cities_boundaries_serdes_tests.cpp]: Done

File: indexer/indexer_tests/cities_boundaries_serdes_tests.cpp (hunk 9 above)
Comment (ygorshenin) [reply indexer/indexer_tests/cities_boundaries_serdes_tests.cpp]: Нет, не много. Это правильная погрешность для координат закодированных в 19 бит.

**Comment by ygorshenin** [comment]
@mpimenov @bykoianko @Zverik PTAL

File: generator/cities_boundaries_builder.cpp [hunk 10]
```diff
@@ -0,0 +1,107 @@
...
+
+#include "indexer/cities_boundaries_serdes.hpp"
+#include "indexer/city_boundary.hpp"
+#include "indexer/classificator.hpp"
+#include "indexer/index.hpp"
+#include "indexer/mwm_set.hpp"
+
+#include "platform/local_country_file.hpp"
+
+#include "coding/compressed_bit_vector.hpp"
```
Comment (mpimenov) [inline generator/cities_boundaries_builder.cpp]: I don't think you need it here.

Comment (ygorshenin) [reply generator/cities_boundaries_builder.cpp]: Done

**Comment by ygorshenin** [comment]
@mpimenov @bykoianko @Zverik PTAL

File: generator/osm_translator.hpp [hunk 11]
```diff
@@ -324,6 +324,15 @@ class OsmToFeatureTranslator
     return true;
   }
 
+  bool IsCityBoundary(FeatureParams const & params)
+  {
+    feature::TypesHolder types;
```
Comment (Zverik) [inline generator/osm_translator.hpp]: Maybe check that the feature is an area? Or call this from inside `IsDrawableLike(AREA)`, like Max suggests.

Comment (ygorshenin) [reply generator/osm_translator.hpp]: But this method is only called from EmitArea, moreover, in the next PR I'll transform it to:
```c++
IsCityOrTown(Types const & types);
```

**Pull request merged**